### ISSUE TITLE
When moving half way in the ligature, the cursor should be half way in the char

### ISSUE THREAD

**Issue by chrilith** [kiliman/operator-mono-lig#35]
Hi,

This is how it is in Fira and sounds more logical than having the cursor in a blank space.

Try for instance to navigate through `>=` in Fira and with Operator-Mono-Lig.

Thanks!

**Comment by kiliman** [comment]
Yes, I agree. I did try to replicate the way Fira Code does it, but was not successful. I'm not sure what trick they used. I noticed they used negative LSB, but they also must include some spacing rules.

I'm not an expert on OpenType, so this is the best I could come up with. Perhaps someone with more experience could point me in the right direction.

Other than that, I hope you're enjoying the ligatures.

Thanks!

**Comment by chrilith** [comment]
You did and still do a great job on that ;)

let's wait to see if someone can help on this!

**Comment by kiliman** [comment]
I've been reading up on how contextual substitutions work in OpenType as well as how it's used in Fira Code. I think I understand it much better now. I'm working on an experimental font with a few ligatures for testing.

Here's an animation of how the spacing works.

![ligature-spacing-test](https://user-images.githubusercontent.com/47168/36437640-8d51af46-1635-11e8-9ad1-494a0aa2b761.gif)

Here's how the glyph is created. 

![image](https://user-images.githubusercontent.com/47168/36437816-11e1cd2c-1636-11e8-8d64-b659b19c8c66.png)

Notice the negative LSB (Left Side Bearing). The way Fira Code works is that instead of substituting `equal` + `equal` into a single double-wide glyph `equal_equal.liga` like I currently do, it actually uses 2 glyphs, one name `LIG` which is simply a single width glyph with no visual representation, plus the `equal_equal.liga` with negative LSB.

So basically `LIG` advances one character width, then the ligature moves back to the left with the negative LSB. The editor and terminal think there are two characters and the caret is displayed properly.

Pretty clever, but makes it a little trickier to design the glyphs because you have to calculate the proper LSB. The actual contextual substitution lookups are really nasty and will have to programmatically generated. I'm working on that now.

**Comment by kiliman** [comment]
I have a working version of the font in VS Code that supports both caret positioning and chained ligatures (#37).

Here's an example:

![ligature-contextual-substitution](https://user-images.githubusercontent.com/47168/36854280-9a7ca4f4-1d3e-11e8-847e-519c50fb2403.gif)

Now to redo all the ligatures 😣 ... I can see the finish line 🏁

**Comment by kiliman** [comment]
I'm almost done with the first set of ligatures. I even included some extra ones for fun.

![image](https://user-images.githubusercontent.com/47168/36905528-9ac30f98-1e01-11e8-86ed-6d0ec64ca056.png)

**Comment by kiliman** [comment]
This has been pushed to the features/contextsubst branch.

**Comment by jabacchetta** [comment]
@kiliman Are the steps still the same for this branch? The `master` branch worked fine, but the build script in the [features/contextsubst](url) branch doesn't appear to be doing anything for me. I'll have to dig deeper when I get the free time.

Might be worthwhile adding a note about this feature in the documentation as well. I just assumed this wasn't possible until I noticed it existed in Fira Code, which led me to this issue.

**Comment by kiliman** [comment]
@jabacchetta currently only the Operator Mono SSm Book and Operator Mono Light (partial) work in the `features/contextsubst` branch. In order to support these features, the ligatures have to be redone.

If you look at the comment above https://github.com/kiliman/operator-mono-lig/issues/35#issuecomment-367047375

You'll see that all the glyphs have to be shifted left with negative left bearing. The original way was to have a character that was 2-3 glyphs wide. So => would be 1250 wide vs 625 for a single glyph.

The new way makes all the ligatures 625 wide, but the extra stuff is shifted to the left. When it is rendered, it actually renders an invisible character LIG followed by the ligature.

So => would be LIG + (=> glyph)... the LIG would advance 625, then the ligature glyph (due to negative left bearing) would actually start rendering over top of the invisible LIG. This is how you get the caret positioning.

Hope that makes sense. I literally had to buy a book on OpenType fonts in order to understand it.

**Comment by dannymcgee** [comment]
Hey @kiliman, would you be willing to share what book you bought and which chapters/sections were relevant for figuring out this issue? I'm trying to make my own original typeface with coding ligatures, and I'm having a hell of a time trying to figure out the maze of substitution/lookup rules to make this work.

**Comment by kiliman** [comment]
@dannymcgee The book is _Fonts & Encodings_ by Yannis Haralambous[1]

Appendix D.9 The OpenType Advanced Typographic Tables (along with a bunch of trial and error) is where I got most of the info.

If you look at the gsub.js file[2] you'll see how I build up the lookup tables. I also talk about it in this comment[3].

I would also recommend this tool: OTMaster Light[4] from Dutch Type Library. It was invaluable in making sure my code was generating the correct tables.

Here's an example showing how the => ligature is rendered. It has checks for repeating characters (so multiple chains of == doesn't render the wrong glyph. Also adds LIG glyphs to support carets inside the ligature glyph.

![subst-sample2](https://user-images.githubusercontent.com/47168/59306672-c7526a80-8c6a-11e9-8880-804977a90048.gif)

[1] https://www.amazon.com/Fonts-Encodings-Advanced-Typography-Everything/dp/0596102429
[2] https://github.com/kiliman/operator-mono-lig/blob/features/contextsubst/gsub.js
[3] https://github.com/kiliman/operator-mono-lig/issues/37#issuecomment-367425462
[4] https://www.fontmaster.nl/index.php/test-demo/

**Comment by kiliman** [comment]
BTW: I only had to dig through all this mess because I couldn't distribute the ligature versions of the fonts directly. 

You can use a font editor like Font Forge or Glyphs to generate the tables for you. Here's the script the FiraCode uses to generated the calt tables.

https://github.com/tonsky/FiraCode/blob/master/clojure/regen_calt.clj
